### PR TITLE
Allow techs to change poster if player option `techCanOverridePoster` is set

### PR DESCRIPTION
**Pull request by LoveIsGrief**
## Description

The option for the player `techCanOverridePoster` is introduced in this commit. It allows techs to update the post whenever they like. `isPosterFromTech_` is introduced as a private player field in order to track when a poster was set by a tech. This allows us to clear the poster whenever the tech is disposed of by the player.

Additionally, attempting to set the same poster more than once will have no effect / no changes will be made, since the poster is the same. This was done in order to stop triggering multiple *posterchange* events when calling `player.poster(aPoster)` with `techCanOverridePoster` set to true.

Related to #4910 

## Specific Changes proposed

 - `techCanOverridePoster` option
 - `isPosterFromTech_` player field
 - `player.poster(aPoster)` triggers a single *posterchange* event
 - `poster.poster(aPoster) && poster.poster(aPoster)` triggers a single *posterchange* event

## Requirements Checklist
- [x] Feature implemented / Bug fixed
- [ ] If necessary, more likely in a feature request than a bug fix
  - [ ] Change has been verified in an actual browser (Chome, Firefox, IE)
  - [x] Unit Tests updated or fixed
  - [ ] Docs/guides updated
  - [ ] Example created ([starter template on JSBin](http://jsbin.com/axedog/edit?html,output))
- [ ] Reviewed by Two Core Contributors

### CODE REVIEW

**Comment by LoveIsGrief** [comment]
I don't know which docs to update or where the options are documented, but I wouldn't mind adding them if someone points me to them.

**Comment by gkatsev** [comment]
For docs, I'd think that [options](https://github.com/videojs/video.js/blob/master/docs/guides/options.md) and [tech](https://github.com/videojs/video.js/blob/master/docs/guides/tech.md) docs get updated. Not sure if there are other things that would need to be updated or not. Updating the docs would be awesome!

File: src/js/player.js [hunk 1]
```diff
@@ -513,6 +516,8 @@ class Player extends Component {
 
     if (this.tech_) {
       this.tech_.dispose();
+      this.isPosterFromTech_ = false;
```
Comment (gkatsev) [inline src/js/player.js]: why only set and remove these inside of this if statement? Maybe we just want to do it unconditionally?

File: src/js/player.js [hunk 2]
```diff
@@ -1017,6 +1022,13 @@ class Player extends Component {
     this.tech_.dispose();
 
     this.tech_ = false;
+
+    if (this.isPosterFromTech_) {
```
Comment (gkatsev) [inline src/js/player.js]: unset the poster if the poster was set via the tech? Interesting. That sounds like a great idea.

File: test/unit/player.test.js [hunk 3]
```diff
@@ -1709,3 +1709,93 @@ QUnit.test('player.duration() sets the value of player.cache_.duration', functio
...
+
+QUnit.test('setPoster in tech with `techCanOverridePoster` in player should override poster', function(assert) {
+  const player = TestHelpers.makePlayer({
+    techCanOverridePoster: true
+  });
+  const posterchangeSpy = sinon.spy();
+  const firstPosterUrl = 'https://wherever.test/test.jpg';
+  const techPosterUrl = 'https://somewhere.text/my/image.png';
+
+  assert.equal(player.options_.techCanOverridePoster, true);
```
Comment (gkatsev) [inline test/unit/player.test.js]: can we add a description to all the assertions?

Comment (LoveIsGrief) [reply test/unit/player.test.js]: Do you mean as comments or here in the PR?

Comment (gkatsev) [reply test/unit/player.test.js]: in the PR, the third argument to the assert calls.

File: src/js/player.js (hunk 1 above)
Comment (LoveIsGrief) [reply src/js/player.js]: I didn't want to be greedy and just wanted to limit the effects of my changes. But I suppose they can unconditionally be set.

I believe a `posterchange` event and an update of the `PosterImage` component would be useless here too. Do you concur?

File: test/unit/player.test.js (hunk 3 above)
Comment (LoveIsGrief) [reply test/unit/player.test.js]: Ah, I see. On it!

File: src/js/player.js (hunk 1 above)
Comment (gkatsev) [reply src/js/player.js]: Yup, no need to do that during cleanup.

File: src/js/player.js [hunk 4]
```diff
@@ -922,7 +927,8 @@ class Player extends Component {
       'poster': this.poster(),
       'language': this.language(),
       'playerElIngest': this.playerElIngest_ || false,
-      'vtt.js': this.options_['vtt.js']
+      'vtt.js': this.options_['vtt.js'],
+      'canOverridePoster': !!this.options_.techCanOverridePoster
```
Comment (gkatsev) [inline src/js/player.js]: seems reasonable.

**Comment by LoveIsGrief** [comment]
Hi, is there something more that needs to be done in order to get this merged?

**Comment by gkatsev** [comment]
Hey, thanks for your patience @LoveIsGrief. I kept it unmerged because we had some fixes that we wanted to merge in before doing another minor release. We're also in the process of doing some tests of this PR as well. I think we can probably get a new minor out this or next week. Nothing more is required on your end, thanks!

**Comment by LoveIsGrief** [comment]
Thanks for the response @gkatsev. I was just curious :smiley: 

Cheers!

**Comment by pagarwal123** [comment]
QA-LGTM

**Comment by gkatsev** [comment]
Thanks @LoveIsGrief!

**Pull request merged**